### ISSUE TITLE
virtual attributes not retrieved using select(),where() and includes()

### ISSUE THREAD

**Issue by jreichert** [rails/rails#15185]
Virtual attributes are properly assigned to the object when combining includes() and select(), but not when combining those with where().  Example:

p = Post.includes(:comments).select("posts.*, 1 as testval")
p.first.title
=> "My Post"
p.first.testval
=> 1

p = Post.includes(:comments).select("posts.*, 1 as testval").where("comments.id < ?",1000).references(:comments)
p.first.title
=> "My Post"
p.first.testval
NoMethodError: undefined method `testval' for #Post:0x007fcd6b93eb68

**Comment by dougc84** [comment]
I have this same issue - Rails 4.1.6, an upgrade to latest 4.1.x doesn't fix anything.

Basically, there's a select statement, with a custom, named attribute, a couple joins, and an includes, just like is mentioned above.  The removal of the includes makes it work but does not return the correct results.  Alternative approaches slow down everything exponentially.

**Comment by p1366** [comment]
We all have the same issue. The issue is in ignoring any columns/aliases when .eager_load is used.

Short and simple fix is to implement support of tN_attrname column aliases.

**Comment by dmgr** [comment]
This is a monkey patch I have prepared to workaround this issue:

```ruby
module ActiveRecord
  class Base
    attr_accessor :_row_
  end

  module Associations
    class JoinDependency
      JoinBase && class JoinPart # :nodoc:
        def instantiate_with_row row, aliases
          model_instance = instantiate_without_row row, aliases
          model_instance._row_ = row
          model_instance
        end; alias_method_chain :instantiate, :row
      end
    end
  end
end
```
Then you can get a column value from row:
```ruby
Post.first._row_['test_val']
```

**Comment by ttrmw** [comment]
@dmgr your monkey patch seems to break eager_load? I get: `ArgumentError Exception: wrong number of arguments (1 for 2)`.

**Comment by Envek** [comment]
I'm suffering from it too. My use case is here: http://stackoverflow.com/questions/37720669/add-computable-column-to-multi-table-select-clause-with-eager-load

And I'd prefer to avoid separate queries (`preload` behavior) as Oracle have limitation of 1000 ids in single query and it's relatively easy to hit it when loading one-to-many associations where many side of relation have tens or hundreds records.

**Comment by Envek** [comment]
@dmgr Thank you! Your monkey patch works fine with ActiveRecord 4.2.6!

**Comment by dmgr** [comment]
@ttrmw I am not sure what exactly the problem may be, but you could try something like this:

```ruby
module ActiveRecord
  class Base
    attr_accessor :_row_
  end

  module Associations
    class JoinDependency
      JoinBase && class JoinPart
        def instantiate_with_row *args
          instantiate_without_row(*args).tap { |i| i._row_ = row }
        end; alias_method_chain :instantiate, :row
      end
    end
  end
end
```

**Comment by alekseyl** [comment]
Hi, everyone. 
I also struck with this problem in my project, also it was impossible to step out of virtual attribute and break it to couple queries also was unacceptable solution. So I came up with my own way of doing it which I moved to a simple gem:

https://github.com/alekseyl/rails_select_on_includes

You are free to use it. It works only for base class now, but if you need some advanced select to child objects, I think I could upgrade it.

**Comment by mswiszcz** [comment]
I wanted to bump this up, I am having the same problems on rails 4.2.7, trying to use sql SUM method aliased as amount_sum, fetch labels for products as well with include, and scoping with several wheres`

**Comment by alekseyl** [comment]
Did you try my monkey-gem?  

https://github.com/alekseyl/rails_select_on_includes

**Comment by mswiszcz** [comment]
@alekseyl Yes, it works perfectly, but I would love to see this fixed in rails. In ruby ecosystem gems have tendency to get abandoned, so I think fixes like this should be incorporated in rails.

Nevertheless, thank you for you work, at the moment it solves the problem 👌

**Comment by oddlyfunctional** [comment]
+1

**Comment by voltechs** [comment]
So, why isn't this fixed in core? Is it intended to be a supported feature?

**Comment by pasivuorio** [comment]
Has someone found a workaround for this problem in Rails 5? rails_select_on_includes gem worked well in Rails 4.2, but when migrating to 5.1.0 it gives following exception:

An ActiveRecord::StatementInvalid occurred in modelname#search:
PG::IndeterminateDatatype: ERROR:  could not determine data type of parameter $4

**Comment by alekseyl** [comment]
Write me an issue on gem, and I'll fix it on upcomming holydays.

**Comment by alekseyl** [comment]
Added rails 5 support for my gem.

**Comment by gduprey** [comment]
So the rails_select_on_includes gem is a brilliant solution -- works perfectly!  

That said, agree with others that if it's this gem can do it, it really should be a supported part of AR when using eager_load. virtual_attributes (in our case, in the form of SQL DB functions) are absolutely critical.  Making multiple round-trip calls to the database server is slow and costly compared to bottling everything you know you need (dependent tables, computed values, DB functions) in a single query.

eager_load is so useful to efficient model use when you know you're going to be needing other tables, I'm not sure why folks think the first solution is to break it into separate queries (especially given the preload method that uses a separate query with a big IN statement is subject to size limits).  I've been doing SQL for the better part of 20 years and inner joins, when dealing with 1-1 associations in a single DB query, is critical to efficient queries.

**Comment by smatyas** [comment]
The [rails_select_on_includes](https://github.com/alekseyl/rails_select_on_includes) patch from @alekseyl works perfectly for me too.
We have to use aggregated and virtual fields in our queries for performance reasons, so from my point of view it should be a must to handle this seamlessly in an ORM.

**Comment by DiegoSalazar** [comment]
I was able to use [rails_select_on_includes](https://github.com/alekseyl/rails_select_on_includes) on Rails 4.1.16 but needed a small patch: https://github.com/alekseyl/rails_select_on_includes/pull/10

**Comment by mengu** [comment]
hi. i believe this must be fixed in AR core. @alekseyl would you be interested in sending a PR?

**Comment by sgrif** [comment]
I'm going to go ahead and close this, as there's been very little movement on it for 3 years. I don't think this is strictly a bug, but I do think this is a gotcha that we can (possibly) improve in a backwards compatible way.

I'm not sure that [my comments from 2014](https://github.com/rails/rails/issues/15185#issuecomment-43549305) still holds or not. That approach is only tenable if we can be certain there is no way to access a column specified in a manual select when used in conjunction with `.includes` and `.references` (e.g. something that triggers column aliasing).

I would still accept a PR that addresses this, if it follows the approach outlined back in my original comments, and if the PR description provides enough information to demonstrate that there is no way this is going to break behavior someone is relying on, or if the PR provides a deprecation cycle and clear migration path for people relying on that behavior.

Looking through the gem linked, that is not an implementation that would be accepted if it were submitted as a PR.

With all of that said, I consider this a gotcha, not a bug -- And I don't see any movement happening any time soon, so I'm going to close this issue.

**Comment by mengu** [comment]
hi @sgrif 

consider the following scenario: we have 3 models: post, comment, author. we get posts with `Post.joins(:author).includes(:author).all`. that's fine. we'd like to get the comment count as well:

`Post.joins(:author).includes(:author).select("(SELECT COUNT(*) FROM comments WHERE comments.post_id = posts.id) AS comment_count").` 

while this generates the correct query, comment_count will not be available in post instances where as it will be available if we don't eager load authors. how is this not a bug?

**Comment by tilhoft** [comment]
I think this is still a Bug, i'm sad about this issue was closed.

**Comment by ncri** [comment]
Yep I agree, just stumbled upon this too.

**Comment by chrishough** [comment]
I just hit this as well, and yes, this is still a bug

**Comment by jstayton** [comment]
If you're OK with one additional query – as opposed to the eager loading behavior of `includes` with `where` – you can use `preload` + `joins` instead, which still allows virtual attributes to be accessed. My guess is that's a totally acceptable trade-off for *most* use-cases. That said, I agree that this is an annoying gotcha that people obviously stumble upon (case in point: this thread).

**Comment by kalinchuk** [comment]
Bug still exists. Any updates?

**Comment by Tarens2** [comment]
If you do queries something like this
`posts = Post.includes(:comments).select("posts.*, 1 as testval").where("posts.active = 1 OR comments.id < 1000").references(:comments)`

you can access custom columns by this code `posts.first.attributes["testval"]`

**Comment by delongGao** [comment]
Still having the same issue with rails 5.2.2 and postgresql. Basically all select statements are ignored when used with `includes` and `references`.

**Comment by joshmn** [comment]
Really kills me. I'm not familiar enough with the underlying parts of ActiveRecord to even attempt at giving it a go, but if someone wants to give me a hint I can take a look.

```
@items = Item.includes(sales_order_items: [:sales_order]).references(sales_order_items: [:sales_order]).where('sales_orders.txn_date > ?', 3.months.ago).select("items.*, sum(sales_order_items.quantity_remaining) as total_quantity").group('sales_order_items.id, sales_orders.id, items.id')
@items.first.total_quantity # undefined method, as expected
```

But if I plug that in `ApplicationRecord.connection.execute(@items.to_sql).to_a.first` and key down, I can see that it is returned, just not mapped.

**Comment by nbrustein** [comment]
Another workaround to this issue is to use `preload` directly instead if `includes`.

So if this doesn't work:

```
MyModel.select("'some_magic' as my_special_property").includes(*some_relationships)
```

this will

```
my_records = MyModel.select("'some_magic' as my_special_property")
ActiveRecord::Associations::Preloader.new.preload(my_records, some_relationships)
```

You're basically forcing activerecord to make multiple queries to load the associations, rather than doing a join. In certain cases, though, this would end up being less efficient, so use with care.

**Comment by junket** [comment]
It would be great to reopen this issue @sgrif :) Thanks @alekseyl for the workaround 👍

**Comment by alible** [comment]
This needs to be fixed in rails I agree.

**Comment by Ardena** [comment]
I'm hitting it too.  Seems like the obvious answer would just be to put the custom columns in the base model's list of attributes.  You could technically put them on all of the eager_loaded models' lists of attributes, though, and that would be fine too.  You could even name the columns something like c_r0 (custom table item 0) so you'd know they were custom.

Definitely feels like a bug, since you ask for something and don't get it available to you.  I don't know how you define "gotcha" differently than bug.

**Comment by dsounded** [comment]
face the same issue

**Comment by fedefa** [comment]
I was at least 1 hour with this until I found a workaround. Just look this:

**Not working**:

Post.includes(:comments).joins(:comments).select("posts.*, 1 as testval")

**Working**:

Post.includes(:comments).joins("JOIN comments ON comments.post_id = posts.id").select("posts.*, 1 as testval")

**Comment by StefSchenkelaars** [comment]
Looks like this bug is a duplicate of #34889 which is fixed in [v6.0.3](https://github.com/rails/rails/blob/v6.0.3/activerecord/CHANGELOG.md#rails-603-may-06-2020) 🎉
